rust/src/fvm/machine.rs

```diff
@@ -31,265 +34,173 @@ lazy_static! {
 /// for some types is due to the generated bindings not liking the
 /// 32bit types as incoming args
 ///
-#[no_mangle]
-#[cfg(not(target_os = "windows"))]
-pub unsafe extern "C" fn fil_create_fvm_machine(
-    fvm_version: fil_FvmRegisteredVersion,
+#[ffi_export]
+fn create_fvm_machine(
+    fvm_version: FvmRegisteredVersion,
     chain_epoch: u64,
     base_fee_hi: u64,
     base_fee_lo: u64,
     base_circ_supply_hi: u64,
     base_circ_supply_lo: u64,
     network_version: u64,
-    state_root_ptr: *const u8,
-    state_root_len: libc::size_t,
-    manifest_cid_ptr: *const u8,
-    manifest_cid_len: libc::size_t,
+    state_root: c_slice::Ref<u8>,
+    manifest_cid: c_slice::Ref<u8>,
     blockstore_id: u64,
     externs_id: u64,
-) -> *mut fil_CreateFvmMachineResponse {
-    catch_panic_response(|| {
-        init_log();
-
-        info!("fil_create_fvm_machine: start");
-
-        let mut response = fil_CreateFvmMachineResponse::default();
-        match fvm_version {
-            fil_FvmRegisteredVersion::V1 => info!("using FVM V1"),
-            //_ => panic!("unsupported FVM Registered Version")
-        }
-
-        let config = Config::default();
-        let chain_epoch = chain_epoch as ChainEpoch;
-
-        let base_circ_supply = TokenAmount::from(
-            ((base_circ_supply_hi as u128) << u64::BITS) | base_circ_supply_lo as u128,
-        );
-        let base_fee =
-            TokenAmount::from(((base_fee_hi as u128) << u64::BITS) | base_fee_lo as u128);
-
-        let network_version = match NetworkVersion::try_from(network_version as u32) {
-            Ok(x) => x,
-            Err(_) => {
-                response.status_code = FCPResponseStatus::FCPUnclassifiedError;
-                response.error_msg =
-                    rust_str_to_c_str(format!("unsupported network version: {}", network_version));
-                return raw_ptr(response);
+) -> repr_c::Box<Result<repr_c::Box<FvmMachine>>> {
+    unsafe {
```

Comment:
Is the `unsafe` block needed here?

Reply:
yes, because it uses an `unsafe` version of the panic catcher